internal/engine/rule_table.go

```diff
@@ -0,0 +1,557 @@
+// Copyright 2021-2024 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package engine
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"sync"
+	"time"
+
+	effectv1 "github.com/cerbos/cerbos/api/genpb/cerbos/effect/v1"
+	policyv1 "github.com/cerbos/cerbos/api/genpb/cerbos/policy/v1"
+	runtimev1 "github.com/cerbos/cerbos/api/genpb/cerbos/runtime/v1"
+	"github.com/cerbos/cerbos/internal/namer"
+	"github.com/cerbos/cerbos/internal/storage"
+	"github.com/cerbos/cerbos/internal/util"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const (
+	storeReloadTimeout = 5 * time.Second
+	storeFetchTimeout  = 2 * time.Second
+)
+
+type RuleTable struct {
+	policyLoader             PolicyLoader
+	log                      *zap.SugaredLogger
+	schemas                  map[string]*policyv1.Schemas
+	policyDerivedRoles       map[string]map[string]*wrappedRunnableDerivedRole
+	scopeMap                 map[string]struct{}
+	scopeScopePermissions    map[string]policyv1.ScopePermissions
+	parentRoles              map[string][]string
+	parentRoleAncestorsCache map[string][]string
+	rules                    []*RuleTableRow
+	mu                       sync.RWMutex
+}
+
+type wrappedRunnableDerivedRole struct {
+	*runtimev1.RunnableDerivedRole
+	constants map[string]any
+}
+
+type RuleTableRow struct {
+	*runtimev1.RuleTable_RuleRow
+	params        *RuleTableRowParams
+	evaluationKey string
+}
+
+type RuleTableRowParams struct {
+	key       string
+	constants map[string]any // conditions can be converted to Go native types at build time
+	variables []*runtimev1.Variable
+}
+
+func NewRuleTable() *RuleTable {
+	return &RuleTable{
+		log:                      zap.S().Named("ruletable"),
+		schemas:                  make(map[string]*policyv1.Schemas),
+		policyDerivedRoles:       make(map[string]map[string]*wrappedRunnableDerivedRole),
+		scopeMap:                 make(map[string]struct{}),
+		scopeScopePermissions:    make(map[string]policyv1.ScopePermissions),
+		parentRoles:              make(map[string][]string),
+		parentRoleAncestorsCache: make(map[string][]string),
+	}
+}
+
+func (rt *RuleTable) WithPolicyLoader(policyLoader PolicyLoader) *RuleTable {
+	rt.policyLoader = policyLoader
+	return rt
+}
+
+func (rt *RuleTable) LoadPolicies(rps []*runtimev1.RunnablePolicySet) {
+	rt.mu.Lock()
+	defer rt.mu.Unlock()
+
+	for _, rp := range rps {
+		rt.addPolicy(rp)
+	}
+}
+
+func (rt *RuleTable) addPolicy(rps *runtimev1.RunnablePolicySet) {
+	switch rps.PolicySet.(type) {
+	case *runtimev1.RunnablePolicySet_ResourcePolicy:
+		rt.addResourcePolicy(rps.GetResourcePolicy())
+	case *runtimev1.RunnablePolicySet_RolePolicy:
+		rt.addRolePolicy(rps.GetRolePolicy())
+	}
+}
+
+func (rt *RuleTable) addResourcePolicy(rrps *runtimev1.RunnableResourcePolicySet) {
```

Comment:
There's a discrepancy between full and partial rule tables. Either this function needs to act on individual `RunnableResourcePolicySet_Policy` instances, or it needs to handle duplicate rows that occur when `GetAll` returns the same policy higher up the scope chain.